transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c

```diff
@@ -0,0 +1,378 @@
+}
+
+void directFile_JNI_OnUnLoad(JNIEnv* env) {
+    // Deleting global refs so their classes can be GCed
+    if (errorInfoConstr != NULL) {
+        (*env)->DeleteGlobalRef(env, (jobject)errorInfoConstr);
+    }
+
+    if (errorInfoClass != NULL) {
+        (*env)->DeleteGlobalRef(env, (jobject)errorInfoClass);
+    }
+}
+
+static inline struct io_control * getIOControl(JNIEnv* env, jobject pointer) {
+    struct io_control * ioControl = (struct io_control *) (*env)->GetDirectBufferAddress(env, pointer);
+    return ioControl;
+}
+
+/**
+ remove an iocb from the pool of IOCBs. Returns null if full
+*/
+static inline struct iocb * getIOCB(struct io_control * control) {
+    struct iocb * iocb = 0;
+
+    pthread_mutex_lock(&(control->iocbLock));
+
+    #ifdef DEBUG
+       fprintf (stderr, "getIOCB::used=%d, queueSize=%d, get=%d, put=%d\n", control->used, control->queueSize, control->iocbGet, control->iocbPut);
+    #endif
+
+    if (control->used < control->queueSize) {
+        control->used++;
+        iocb = control->iocb[control->iocbGet++];
+
+        if (control->iocbGet >= control->queueSize) {
+           control->iocbGet = 0;
+        }
+    } else {
+        #ifdef DEBUG
+            fprintf (stderr, "Could not find iocb\n");
+        #endif
+    }
+
+    pthread_mutex_unlock(&(control->iocbLock));
+    return iocb;
+}
+
+/**
+  Put an iocb back on the pool of IOCBs
+*/
+static inline void putIOCB(struct io_control * control, struct iocb * iocbBack) {
+    pthread_mutex_lock(&(control->iocbLock));
+
+    #ifdef DEBUG
+       fprintf (stderr, "putIOCB::used=%d, queueSize=%d, get=%d, put=%d\n", control->used, control->queueSize, control->iocbGet, control->iocbPut);
+    #endif
+
+    control->used--;
+    control->iocb[control->iocbPut++] = iocbBack;
+    if (control->iocbPut >= control->queueSize) {
+       control->iocbPut = 0;
+    }
+    pthread_mutex_unlock(&(control->iocbLock));
+}
+
+static inline void * getBuffer(JNIEnv* env, jobject pointer) {
+    return (*env)->GetDirectBufferAddress(env, pointer);
+}
+
+/**
+  Everything that is allocated here will be freed at deleteContext when the class is unloaded.
+*/
+JNIEXPORT jobject JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_newContext(JNIEnv* env, jclass clazz, jint queueSize) {
+    io_context_t libaioContext;
+    int i = 0;
+
+    int res = io_queue_init(queueSize, &libaioContext);
+    if (res) {
+        // Error, so need to release whatever was done before
+        free(libaioContext);
+
+        throwRuntimeExceptionErrorNo(env, "Cannot initialize queue:", res);
+        return NULL;
+    }
+    struct io_control * theControl = (struct io_control *) malloc(sizeof(struct io_control));
+    if (theControl == NULL) {
+        throwOutOfMemoryError(env);
+        return;
+    }
+
+    res = pthread_mutex_init(&(theControl->iocbLock), 0);
+    if (res)
+    {
+        throwRuntimeExceptionErrorNo(env, "Can't initialize mutext:", res);
+        return;
```

Comment:
Should we ensure we cleanup `free(libaioContext);` an other memory that is cleaned up above here?